Add tensor.{adjoint(),H,mT,mH} methods and properties

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64181
* #64180
* __->__ #64179

This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

Differential Revision: [D30730483](https://our.internmc.facebook.com/intern/diff/D30730483)

lezcano committed:
Add tensor.{adjoint(),H,mT,mH} methods and properties

This PR follows the discussion in
https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

[ghstack-poisoned]

Lezcano commented:
@ezyang gh added you as code owner, but could you still have a look at this one?

anjali411 commented on docs/source/tensors.rst:
```diff
@@ -177,6 +177,9 @@ Tensor class reference
 
 .. autoattribute:: Tensor.T
+.. autoattribute:: Tensor.H
+.. autoattribute:: Tensor.mT
+.. autoattribute:: Tensor.mH
```
why do we want these attributes?

Lezcano replied:
See the OP: https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

anjali411 reviewed (commented):
missing tests but looks good otherwise

Lezcano commented:
Do you know where should the tests go? I was not able to find the tests for `.T`.
Also, `mypy` seems to complain that `error: "Tensor" has no attribute "mT" `. How can I tackle that one? Perhaps @rgommers knows?

Lezcano commented:
It doesn't work, and it doesn't work for `a.T` in `master` either.

anjali411 commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -28,10 +28,19 @@
+    'view_as_real',
+    '_conj',
+    '_neg_view',
+    'matrix_H',
+    'mH',
```
you probably don't need these since they are composite ops and we wouldn't wanna replay the op if these functions are called on real tensors

Lezcano replied:
Should I leave `adjoint`?

Lezcano replied:
I assume from the discussion below that I should also delete `adjoint`.

anjali411 commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -28,10 +28,19 @@
 
 VIEW_FUNCTIONS = {
     'numpy_T': 'self',
+    'mT': 'self',
```
again same not sure if you need this since it's a composite op cc. @albanD

anjali411 replied:
synced offline -- @Lezcano since all these functions are composite ops they don't need to be added in `VIEW_FUNCTIONS` or `RETURNS_VIEWS_OF_INPUT `, instead we only need entries in `RETURNS_VIEWS_OF_INPUT` below

anjali411 commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -2178,6 +2178,31 @@ Tensor numpy_T(const Tensor &self) {
 }
 
+Tensor matrix_H(const Tensor &self) {
+  TORCH_CHECK(self.dim() == 2,
```
could you add tests to assert these errors for each of the tensor attributes?

anjali411 commented on torch/csrc/jit/frontend/builtin_functions.cpp:
```diff
@@ -49,6 +49,12 @@ def ndim(a : Tensor) -> int:
   return a.dim()
 def T(a : Tensor) -> Tensor:
   return a.numpy_T()
+def H(a : Tensor) -> Tensor:
```
@SplitInfinity could you help review this?

anjali411 replied:
cc. @eellison

anjali411 replied:
should we add entries here: https://github.com/pytorch/pytorch/blob/ac99d63f83ceaee4a95e7baa8a52fba09903d00b/torch/csrc/jit/frontend/sugared_value.cpp#L119 ?

SplitInfinity replied:
Looks good!

SplitInfinity replied:
And yes I think you need the changed to `sugared_value.cpp` so that the JIT compiler can recognize these new properties.

anjali411 commented on torch/csrc/jit/frontend/builtin_functions.cpp:
```diff
@@ -49,6 +49,12 @@ def ndim(a : Tensor) -> int:
   return a.numpy_T()
+def H(a : Tensor) -> Tensor:
+  return a.matrix_H()
+def mT(a : Tensor) -> Tensor:
```
these aren't tensor attributes right? at the least, these shouldn't be added here

SplitInfinity replied:
I thought they were properties? As per the PR title?

ezyang commented:
@anjali411 I'm going to let you take point on this, including landing, is that OK?

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

anjali411 commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -28,7 +28,12 @@
 # A map: function name => name of the argument that all outputs are view of
 
-VIEW_FUNCTIONS_WITH_METADATA_CHANGE = ['view_as_complex', 'view_as_real', '_conj', '_neg_view']
+VIEW_FUNCTIONS_WITH_METADATA_CHANGE = [
```
@Lezcano can you please update RETURNS_VIEWS_OF_INPUT to include the newly added views?

anjali411 replied:
`matrix_H, mT, mH, adjoint`

albanD replied:
That sounds like the right thing to do.

Lezcano replied:
By the description in the file, it seems like `numpy_T` should also be in `RETURNS_VIEWS_OF_INPUT` rather than in `RETURNS_VIEW`...

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

anjali411 commented on torch/csrc/jit/passes/normalize_ops.cpp:
```diff
@@ -127,6 +127,7 @@ const std::unordered_map<Symbol, Symbol>& getOperatorAliasMap() {
       {aten::special_xlogy, aten::xlogy},
       {aten::special_log_softmax, aten::log_softmax},
       {aten::orgqr, aten::linalg_householder_product},
+      {aten::adjoint, aten::mH},
```
why wasn't `mT` included here?

Lezcano replied:
because `mT` is not an alias of any other function.

mruberry replied:
"adjoint" might be the more readable term than "mH" but whatever seems fine

Lezcano replied:
One problem of doing it that way is that I cannot add `mH` as an alias in the `OpInfo`. Leaving it like this for now.

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

rgommers commented:
> Also, `mypy` seems to complain that `error: "Tensor" has no attribute "mT" `. How can I tackle that one?

Add the annotation in `torch/_C/__init__.pyi.in`:
```python
class _TensorBase(metaclass=_TensorMeta):
    ...
    T: Tensor
    mT: Tensor
```

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

Lezcano commented:
Thank you @rgommers ! Fixed.

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30730483).

eellison commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8656,6 +8668,43 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+           dtypesIfCUDA=all_types_and_complex_and(torch.bool, torch.bfloat16, torch.half),
+           supports_out=False,
+           supports_forward_ad=True,
+           skips=(  # Fails due to the lambda wrapper
```
What is the error that you are seeing? It should just skip the lambda for scripting: https://github.com/pytorch/pytorch/blob/master/test/test_ops.py#L720

Lezcano replied:
```python
________________________ TestJitMETA.test_variant_consistency_jit_mT_meta_float32 ________________________                                                                                                           
Traceback (most recent call last):                                                                                                                                                                                   
  File "/home/lezcano/git/pytorch/pytorch/test/test_ops.py", line 799, in test_variant_consistency_jit                                                                                                               
    assert tested, "JIT Test does not execute any logic"
AssertionError: JIT Test does not execute any logic
assert False                                         
_______________________ TestJitCUDA.test_variant_consistency_jit_mT_cuda_complex64 _______________________
Traceback (most recent call last):
  File "/home/lezcano/git/pytorch/pytorch/test/test_ops.py", line 799, in test_variant_consistency_jit
    assert tested, "JIT Test does not execute any logic"
AssertionError: JIT Test does not execute any logic
assert False                                         
________________________ TestJitCUDA.test_variant_consistency_jit_mT_cuda_float32 ________________________
Traceback (most recent call last):
  File "/home/lezcano/git/pytorch/pytorch/test/test_ops.py", line 799, in test_variant_consistency_jit
    assert tested, "JIT Test does not execute any logic"
AssertionError: JIT Test does not execute any logic
assert False  
```

eellison commented on torch/csrc/jit/frontend/builtin_functions.cpp:
```diff
@@ -49,6 +49,12 @@ def ndim(a : Tensor) -> int:
+def mT(a : Tensor) -> Tensor:
+  return a.mT
+def mH(a : Tensor) -> Tensor:
+  return a.mH
```
can we add some tests for these ? When you add properties to tensors it does not manually generate any operators. I think you need to add a `prim::mH` operator to https://github.com/pytorch/pytorch/blob/7fff21afcf2f62d1a44ab8c5c5aebf25174b7d30/torch/csrc/jit/runtime/register_prim_ops.cpp. 


See: https://github.com/pytorch/pytorch/issues/30775 although I don't think that should be closed..

Lezcano replied:
Where should these tests go and how should they look like?
Also, I did not see an entry in that table for `T`, how should the entries for these operations look like? I do not know anything about the internal workings of JIT, sorry.

eellison reviewed (commented):
couple questions..

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -2181,6 +2181,43 @@ Tensor numpy_T(const Tensor &self) {
+Tensor _adjoint(const Tensor &self, const bool transpose, const char* const name) {
+  const auto ndim = self.dim();
+  TORCH_CHECK(ndim != 1,
+      "tensor.", name, " only supported on matrices or batches of matrices. Got 1-D tensor.");
```
"is only"

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -2181,6 +2181,43 @@ Tensor numpy_T(const Tensor &self) {
+Tensor matrix_H(const Tensor &self) {
+  const auto ndim = self.dim();
+  TORCH_CHECK(ndim == 2 || ndim == 0,
+      "tensor.H only supported on matrices (2-D tensors). Got ", ndim, "-D tensor.",
```
"is only"

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3262,6 +3262,21 @@
 - func: numpy_T(Tensor(a) self) -> Tensor(a)
   variants: method
 
+# Exposed on Python as an attribute 'H'
```
Nice comments

mruberry commented on torch/_torch_docs.py:
```diff
@@ -435,6 +435,13 @@ def merge_dicts(*dicts):
             [ 0.7573, -3.9555, -2.8681]])
 """.format(**common_args, **tf32_notes))
 
+add_docstr(torch.adjoint,
```
For consistency, the method variant typically points to the function variant and it'd be nice to maintain that here to save people searching for this op a click

mruberry replied:
From a user's perspective, mH the property is equivalent to calling the function adjoint, right? I'm not sure I'd bother describing one as an alias of another

Lezcano replied:
What are you suggesting in the second message?

mruberry replied:
For .mH it could say something straightforward like "Accessing this property is equivalent to calling torch.adjoint(t)."

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8656,6 +8668,43 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+               SkipInfo("TestJit", "test_variant_consistency_jit"),),
+           sample_inputs_func=sample_inputs_T),
+    OpInfo('mT',
+           op=lambda x: x.mT,
```
Nice OpInfos

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -4845,6 +4845,49 @@ def callable(a, b) -> number
+               r"""
+Returns a view of a matrix (2-D tensor) conjugated and transposed.
+
+``x.H`` is equivalent to ``x.transpose(0, 1).conj()`` for complex matrices and
```
Great description

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -4845,6 +4845,49 @@ def callable(a, b) -> number
+
+add_docstr_all('mT',
+               r"""
+Is this tensor with the last two dimensions transposed.
```
"Returns a view of its tensor with the last two dimensions transposed"?

Lezcano replied:
I just went with the previous wording that `.T` had, but I admit it's quite weird. I'll update it to yours.

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -4845,6 +4845,49 @@ def callable(a, b) -> number
+
+add_docstr_all('adjoint',
+               r"""
+adjoint(Tensor) -> Tensor
```
index_fill() -> Tensor?

(this is part of why documenting the function is easier)

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -4845,6 +4845,49 @@ def callable(a, b) -> number
+adjoint(Tensor) -> Tensor
+Returns a view of the tensor conjugated and with the last two dimensions transposed.
+
+``x.adjoint()`` is equivalent to ``x.transpose(-2, -1).conj()`` for complex tensors and
```
Cool

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -4845,6 +4845,49 @@ def callable(a, b) -> number
+
+Example::
+    >>> x = torch.arange(4, dtype=torch.float)
+    >>> A = torch.complex(x, x.flip(0)).reshape(2, 2)
```
starting with a floating point tensor here seems a little circuitous but it's probably fine; might be easier to just construct the complex tensor by hand to being so people don't think the flipping and reshaping are relevant to what the op does

Lezcano replied:
I liked this idea so that it's reproducible. This operation is just moving stuff around and negating some parts. I liked a simple example because it's easier to follow what goes where. I've simplified it a bit more even, tell me what you think.

mruberry reviewed (commented):
Hey @Lezcano!

Cool stuff. I made a few comments inline for you review. @anjali411 is still this PR's principal shepherd.

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

Differential Revision: [D30730483](https://our.internmc.facebook.com/intern/diff/D30730483)

[ghstack-poisoned]

Lezcano commented:
@mruberry addressed your comments. Once we sort out the JIT issues with @eellison , this should be ready to go.

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

Differential Revision: [D30730483](https://our.internmc.facebook.com/intern/diff/D30730483)

[ghstack-poisoned]

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

Differential Revision: [D30730483](https://our.internmc.facebook.com/intern/diff/D30730483)

[ghstack-poisoned]

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -4839,12 +4839,43 @@ def callable(a, b) -> number
+               r"""
+adjoint() -> Tensor
+
+Alias for :func:`adjoint`
```
For consistency with other methods this would be "See :func:\`torch.adjoint\`"

Lezcano replied:
A quick `git grep` shows that both expressions are used across the docs. I'll leave as it is for now, and then I can submit a PR making all the uniform via a regex. Which of the two you think it's better to leave?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -435,6 +435,25 @@ def merge_dicts(*dicts):
+``x.transpose(-2, -1)`` for real tensors.
+
+Example::
+    >>> x = torch.arange(4, dtype=torch.float)
```
Cool example

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

Differential Revision: [D30730483](https://our.internmc.facebook.com/intern/diff/D30730483)

[ghstack-poisoned]

Lezcano commented:
This PR is still missing ONNX support (I don't know how to add this) and the JIT issues mentioned above + tests for JIT.

mruberry commented:
> This PR is still missing ONNX support (I don't know how to add this) and the JIT issues mentioned above + tests for JIT.

Updating all of them would be great!

The correct language would be something like "The method variant of `foo`", I think. Because the method "foo" is not an alias for the function "foo," which would imply that `Tensor.foo is `foo`, which is not correct.

anjali411 commented:
> This PR is still missing ONNX support (I don't know how to add this) 

https://pytorch.org/docs/master/onnx.html#adding-support-for-operators

mruberry commented:
> This PR is still missing ONNX support (I don't know how to add this) and the JIT issues mentioned above + tests for JIT.

I also don't know if ONNX support is required for this PR. Maybe there's something special about properties that will break things if we don't add it, but for a typical op we don't add ONNX support when adding the operator. It can be added in a follow-up PR or (more often) if someone requests ONNX add it.

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

Differential Revision: [D30730483](https://our.internmc.facebook.com/intern/diff/D30730483)

[ghstack-poisoned]

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

Differential Revision: [D30730483](https://our.internmc.facebook.com/intern/diff/D30730483)

[ghstack-poisoned]

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

Differential Revision: [D30730483](https://our.internmc.facebook.com/intern/diff/D30730483)

[ghstack-poisoned]

anjali411 commented on test/test_jit.py:
```diff
@@ -710,6 +710,44 @@ def check(x, y):
+            return torch.equal(x.mT, x.transpose(-2, -1))
+
+        x = torch.rand(3, 4)
+        self.assertTrue(check(x))
```
I think we should use `checkScript` instead of `torch.jit.script`. It verifies that a given script generates the same output as the Python version using the given inputs. So we don't need any additional equality check since we already have correctness checks at other places

```
def test_T_mT_H_mH(self):
    def T(x):
        return x.mT
    def mT(x):
        return x.mT
    def H(x):
        return x.H
    def mH(x):
        return x.mH
    
    x = torch.rand(3, 4)
    y = make_tensor((3, 4), device="cpu", dtype=torch.complex64)
    
    self.checkScript(T, (x, ))
    self.checkScript(mT, (x, ))
    self.checkScript(H, (x, ))
    self.checkScript(mH, (x, ))
    self.checkScript(T, (y, ))
    self.checkScript(mT, (y, ))
    self.checkScript(H, (y, ))
    self.checkScript(mH, (y, ))
```

Lezcano replied:
Added.

anjali411 replied:
I don't see it in the PR. did you push your local changes yet?

anjali411 commented:
Test failures on the PR look real:
```
======================================================================
ERROR [0.006s]: test_jit_alias_remapping_mH_cpu_float32 (__main__.TestJitCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 373, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 368, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 734, in test_wrapper
    return test(*args, **kwargs)
  File "test_ops.py", line 908, in test_jit_alias_remapping
    FileCheck().check(op.aten_name).check_not(variant_name).run(graph)
RuntimeError: Expected to find "mH" but did not find it
Searched string:
graph(%t0.1 : Tensor):
~~ <--- HERE
  %1 : Tensor = aten::adjoint(%t0.1) # <string>:3:35
  return (%1)
From CHECK: mH
```

@eellison can probably help suggest a fix

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

Differential Revision: [D30730483](https://our.internmc.facebook.com/intern/diff/D30730483)

[ghstack-poisoned]

Lezcano commented:
@anjali411 Added now and the CI problems should be gone.
Waiting for a review from @tugsbayasgalan or @eellison on the JIT part, and this should be good to go.

Lezcano commented:
@anjali411 @tugsbayasgalan I have no idea if the mobile / JIT errors are related to this PR or not really.

tugsbayasgalan commented:
@Lezcano do you still see the above error or has it been resolved?

Lezcano commented:
@tugsbayasgalan I fixed that error. There are some errors in the CI, but I don't know if they are related to this PR. Also, could you do a full review of the parts that involve JIT in this PR? Thanks!

tugsbayasgalan commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -116,6 +116,38 @@ static const OperatorGeneratorArgs opGenArgs[] = {
         },
         aliasAnalysisFromSchema()),
+    OperatorGeneratorArgs(
+        TORCH_SELECTIVE_SCHEMA("aten::numpy_T(Tensor(a) self) -> Tensor(a)"),
```
could you make this a overload method since numpy_T is already exist in native_functions.yaml. For example, you can take a look at aten::slice.str or aten::slice.list. Our schema matching logic should be able to automatically figure out which one you are using. 
cc: @eellison

Lezcano replied:
You mean for this one or for the 4 ops? All the ops are in `aten::` so I assume that the treatment should be the same for all of them.

Also, I do not know how should I write it. I would appreciate it if you could write this bit for me, as I do not know anything about the JIT part really.

anjali411 replied:
> could you make this a overload method since numpy_T is already exist in native_functions.yaml

@tugsbayasgalan why do we need an entry for these ops here given that they are already aten functions?

tugsbayasgalan replied:
I meant something like aten::numpy_T.<something> should work.

tugsbayasgalan reviewed (commented):
Overall looks pretty good! Let me know if the test failure doesn't go away after adding overload

Lezcano commented on torch/csrc/jit/frontend/sugared_value.cpp:
```diff
@@ -108,18 +108,33 @@ std::shared_ptr<SugaredValue> SimpleValue::attr(
+           {"is_leaf", "aten"},
+           {"requires_grad", "prim"},
+           {"layout", "prim"},
+           {"T", "prim"},
+           {"H", "prim"},
+           {"mT", "aten"},
+           {"mH", "aten"},
```
@tugsbayasgalan Is this correct? I did not know whether they should be `"prim"` or `"aten"` or what really.

tugsbayasgalan replied:
I don;t think it matters. But i think prim would be better

Lezcano replied:
As per https://github.com/pytorch/pytorch/blob/50edc2679d53b38ab4fc8ab684cd0f8f50df8358/aten/src/ATen/core/interned_strings.h#L502-L505
I believe that it's correct as it stands, as `at::mT` and `at::mH` do exist, while `at::T` and `at::H` do not (they are called `at::numpy_T` and `at::matrix_H`).
Is this correct?

tugsbayasgalan replied:
So I discussed with @eellison and we concluded that it doesn't really matter what the namespace is. But I said prim is better because these newly added operator names can collide with the entries in native_functions.yaml. So I think maybe they all should be prim? @eellison let us know if it is wrong

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

Differential Revision: [D30730483](https://our.internmc.facebook.com/intern/diff/D30730483)

[ghstack-poisoned]

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

Differential Revision: [D30730483](https://our.internmc.facebook.com/intern/diff/D30730483)

[ghstack-poisoned]

tugsbayasgalan commented:
@Lezcano is there anything left from JIT side? Looks like all tests are passing

Lezcano commented:
@tugsbayasgalan All good now, thank you. Just waiting for @anjali411 to get her new laptop so that she's able to land this. Thanks!

lezcano committed:
Update on "Add tensor.{adjoint(),H,mT,mH} methods and properties"


This PR follows the discussion in https://github.com/pytorch/pytorch/issues/45063#issuecomment-904431478

Fixes https://github.com/pytorch/pytorch/issues/45063

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

Differential Revision: [D30730483](https://our.internmc.facebook.com/intern/diff/D30730483)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30730483).
